### PR TITLE
Specify additional initialization functions to prepack

### PR DESCRIPTION
**Pull request by hermanventer**
Provide a new option to specify a number of additional functions to prepack (in addition to the global code "function", that is). These functions need to exist in the global scope, have no parameters, be well behaved (never throw an exception) and should not depend on each other (the order in which they get called should not matter).

At the moment there are checks only for existence, good behavior and no write-write conflicts. Also, the functions are not actually prepacked yet.

### CODE REVIEW

File: src/methods/join.js [hunk 1]
```diff
@@ -589,3 +589,44 @@ export function joinBooleans(v1: void | boolean, v2: void | boolean): void | boo
...
+    if (!m2.has(key)) return;
+    let val2 = m2.get(key);
+    let val3 = intersect(key, val1, val2);
+    m3.set(key, val3);
+  });
+  return m3;
+}
+
+export function intersectEffects(realm: Realm, e1: Effects, e2: Effects): Effects {
+  let [c1, g1, b1, p1, o1] = e1;
```
Comment (NTillmann) [inline src/methods/join.js]: Named effect properties would be great. Very mysterious...

File: src/methods/join.js [hunk 2]
```diff
@@ -589,3 +589,44 @@ export function joinBooleans(v1: void | boolean, v2: void | boolean): void | boo
...
+  let eqVal = (v1, v2) =>
+    v1 === v2 ||
+    (v1 instanceof ConcreteValue && v2 instanceof ConcreteValue && StrictEqualityComparison(realm, v1, v2));
+  let eqDesc = (d1, d2) => d1 && d2 && equalDescriptors(d1, d2) && eqVal(d1.value, d2.value);
+  let p3 = intersectMaps(p1, p2, (b, d1, d2) => (eqDesc(d1, d2) ? d1 : { value: conflict }));
+  // ignore created objects
+  o1;
+  o2;
+
+  return [realm.intrinsics.empty, new Generator(realm), b3, p3, new Set()];
```
Comment (NTillmann) [inline src/methods/join.js]: Judging by the function name `intersectEffects`, I find `realm.intrinsics.empty` and perhaps also `new Generator(realm)` surprising. Expected would be to at least check if c1 === c2, and then return c1 else empty, or probably better, generating code that does it dynamically. Similarly, deep equality check for generator.

I understand that this isn't needed for the concrete scenario at hand, but at least one prominent comment would be good.

File: src/options.js [hunk 3]
```diff
@@ -111,4 +114,6 @@ export function getSerializerOptions({
     singlePass,
     trace,
   };
+  if (additionalFunctions) result.additionalFunctions = additionalFunctions;
```
Comment (NTillmann) [inline src/options.js]: Why this special treatment? Am I missing some Flow subtlety?

File: src/serializer/functions.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+export class Functions {
+  constructor(realm: Realm, functions: ?Array<string>) {
+    this.realm = realm;
+    this.functions = functions;
+  }
+
+  realm: Realm;
+  functions: ?Array<string>;
+
+  evaluateNodeForEffects(node: BabelNode) {
```
Comment (NTillmann) [inline src/serializer/functions.js]: This seems like a generally useful function. I can see how we can cut a few lines e.g. in modules.js by using this.

File: src/methods/join.js (hunk 1 above)
Comment (hermanventer) [inline src/methods/join.js]: Yes, but out of scope for this request.

File: src/options.js (hunk 3 above)
Comment (hermanventer) [inline src/options.js]: additionalFunctions, like all other option properties, is optional. If present, however, its value must be an array. This is the only way I could find that would satisfy these two conditions at once.

File: src/serializer/functions.js (hunk 4 above)
Comment (hermanventer) [inline src/serializer/functions.js]: Will put it in Realm.

File: src/methods/join.js (hunk 2 above)
Comment (hermanventer) [inline src/methods/join.js]: I've changed the name to strongly suggest that only bindings get intersected.

**Pull request closed without merging**